typescript/utils/src/yaml.ts

```diff
@@ -15,3 +22,406 @@ export function tryParseJsonOrYaml<T = any>(input: string): Result<T> {
+}
+
+/**
+ * Inserts comment blocks before a line if they match
+ */
+function insertCommentsBeforeLine(
+  lineNum: LineIndex,
+  commentBlocks: Map<LineIndex, CommentBlock>,
+  originalLines: string[],
+  transformedContentMap: Map<string, LineIndex>,
+  indentationMap: Map<LineIndex, string>,
+  usedComments: Set<string>,
+  result: string[],
+): void {
+  for (const [commentLine, comments] of commentBlocks.entries()) {
+    const bestMatch = findBestMatchForComment(
+      commentLine - comments.length,
+      originalLines,
+      transformedContentMap,
+    );
+
+    if (bestMatch === lineNum) {
+      // Add comments with proper indentation
+      const indentation = indentationMap.get(lineNum) || '';
+      for (const comment of comments) {
+        if (!usedComments.has(comment)) {
+          result.push(
+            comment.startsWith('#') ? indentation + comment.trim() : comment,
+          );
+          usedComments.add(comment);
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Adds a content line with its inline comment if it has one
+ */
+function addContentLineWithInlineComment(
+  line: string,
+  trimmedLine: string,
+  contentToLineMap: Map<string, LineIndex>,
+  inlineCommentMap: Map<LineIndex, string>,
+  result: string[],
+): void {
+  const contentWithoutComment = trimmedLine.trim();
+  const originalLineNum = contentToLineMap.get(contentWithoutComment);
+
+  if (originalLineNum && inlineCommentMap.has(originalLineNum)) {
+    // Has inline comment - reattach it
+    const inlineComment = inlineCommentMap.get(originalLineNum);
+    result.push(`${line} ${inlineComment}`);
+  } else {
+    result.push(line);
+  }
+}
+
+/**
+ * Transforms YAML content by applying a custom transformer function while preserving comments.
+ *
+ * @param content - Original YAML content as a string
+ * @param transformer - A function that transforms the parsed YAML data
+ * @returns The transformed YAML content as a string with comments preserved
+ */
+export function transformYaml<T extends YamlNode = YamlNode>(
+  content: string,
+  transformer: (data: T) => T,
+): string {
+  const parsedDoc = parseDocument(content, { keepSourceTokens: true });
+  const newDoc = new Document();
+  newDoc.contents = transformer(parsedDoc.toJSON());
+
+  return preserveYamlComments(content, newDoc.toString());
+}
+
+export type ArraySortConfig = {
+  arrays: Array<{
+    path: string;
+    sortKey: string;
+  }>;
+};
+
+function findSortKeyForPath(
+  path: string[],
+  config: ArraySortConfig,
+): string | null {
+  const matchingConfig = config.arrays.find(({ path: configPath }) => {
+    const patternParts = configPath.split('.');
+
+    if (path.length !== patternParts.length) {
+      return false;
+    }
+
+    return path.every((part, idx) => {
+      const patternPart = patternParts[idx];
+      if (patternPart === '*') return true;
+      if (patternPart.endsWith('[]')) return patternPart.slice(0, -2) === part;
+      return patternPart === part;
```

Comment:
What is the purpose of a pattern like `foo[].bar`? Or does it work as intended
Will this handle a path like `foo.4.bar`?

It seems we may need tests for this algorithm.